Use llm provided lines

lemaxw opened the pull request:
- Added option to use several llm models (like --llm chatgpt grok), so it will use it
- Added metrics: tokens, count of bugs found, the info published as additional general review

lemaxw committed:
fixed small issues

lemaxw committed:
added try with several LLMs
counts token, and prints amount of bugs

lemaxw committed:
Merge remote-tracking branch 'upstream/use-llm-provided-lines' into use-llm-provided-lines

lemaxw committed:
Merge remote-tracking branch 'upstream/json-response' into use-llm-provided-lines

pavel-anymaint commented on CHANGES.md:
```diff
@@ -14,6 +13,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 - Introduced reviewer class, which creates prompts
 - LLM output in JSON
 - Issues lines detection based on LLM output only
+- Added option to use several llm models (like --llm chatgpt grok), so it will use it
```
`so it will use it` - improve language pls

pavel-anymaint commented on chatgpt_llm.py:
```diff
@@ -29,8 +29,14 @@ def answer(self, system_prompt: str, user_prompt: str, content: str) -> str:
             )
             raw_response = response.choices[0].message.content.strip()
+            usage = response.usage
+            if usage:
```
Print usage in one place, not in every LLM and in log.debug. Not interesting to most people most of the time

pavel-anymaint commented on grok_llm.py:
```diff
@@ -40,13 +40,17 @@ def answer(self, system_prompt: str, user_prompt: str, content: str) -> str:
             logging.debug(f"Raw Response:\n{raw_response[:LOG_CHAR_LIMIT]}... (truncated)")
-            return raw_response
+            usage = result.get("usage")
+            if usage:
```
Print usage in one place, not in every LLM and in log.debug. Not interesting to most people most of the time

pavel-anymaint commented on llm_code_reviewer.py:
```diff
@@ -74,16 +77,17 @@ def review_pr(self, pr: Any, repository: str, pr_number: int) -> LLMReviewResult
             content=content
         )
 
+        if raw_response:
```
Rename variable, it is not raw response anymore

pavel-anymaint commented on models.py:
```diff
@@ -30,30 +42,108 @@ def from_dict(cls, data: dict) -> 'CodeReview':
+    def summarize_reviews(self, reviews: List[CodeReview], total_tokens: int, 
+                prompt_tokens: int, completion_tokens: int) -> Dict[str, int]:
+        totals = {
+            "bug_count": 0,
```
Not sure we need all kinds of messages. I'll test it, but I suspect it will give too many mistakes

lemaxw replied:
This is severity-based indicators: Categorizes outputs by severity, allowing developers to decide if an issue merits later examination. I can hide it under parameter if you wish...

lemaxw committed:
fixes following PR requests

lemaxw committed:
fixed according request changes

Commit:
put print statistic under a flag --add_statistic_info
filter out from full context new and deleted files
limited to not pass too long files to LLM
limited general prompt to 500K

pavel-anymaint commented on chatgpt_llm.py:
```diff
@@ -29,8 +29,12 @@ def answer(self, system_prompt: str, user_prompt: str, content: str) -> str:
                 temperature=0.0,
```
llm_interface and all implementations define str as return type: answer(self, system_prompt: str, user_prompt: str, content: str) -> str:

pavel-anymaint commented on gemini_llm.py:
```diff
@@ -27,8 +27,12 @@ def answer(self, system_prompt: str, user_prompt: str, content: str) -> str:
             raw_response = response.text.strip()
-            logging.debug(f"Raw Response:\n{raw_response[:LOG_CHAR_LIMIT]}... (truncated)")
-            return raw_response
+            usage = response.usage_metadata            
```
Did you test it? ("'GenerateContentResponse' object has no attribute 'usage_metadata'",)

lemaxw replied:
according to the api, it should have it:
https://googleapis.github.io/python-genai/genai.html#genai.types.GenerateContentResponse.usage_metadata

however I don't use gemini api, so couldn't test it...

lemaxw committed:
fixed interface errors
